src/repositories/updater.rs

```diff
@@ -0,0 +1,366 @@
+            updaters.push(Box::new(updater));
+        }
+        if let Ok(updater) = GitLab::new("gitlab.freedesktop.org", &None) {
+            updaters.push(Box::new(updater));
+        }
+        Self { updaters, pool }
+    }
+
+    pub(crate) fn load_repository(&self, metadata: &MetadataPackage) -> Result<Option<i32>> {
+        let url = match &metadata.repository {
+            Some(url) => url,
+            None => {
+                debug!("did not collect stats as no repository URL was present");
+                return Ok(None);
+            }
+        };
+        let mut conn = self.pool.get()?;
+        self.load_repository_inner(&mut conn, url)
+    }
+
+    fn load_repository_inner(&self, conn: &mut Client, url: &str) -> Result<Option<i32>> {
+        let name = match repository_name(url) {
+            Some(name) => name,
+            None => return Ok(None),
+        };
+
+        // Avoid querying the APIs for repositories we already loaded.
+        if let Some(row) = conn.query_opt(
+            "SELECT id FROM repositories WHERE name = $1 AND host = $2 LIMIT 1;",
+            &[&format!("{}/{}", name.owner, name.repo), &name.host],
+        )? {
+            return Ok(Some(row.get("id")));
+        }
+        for updater in self.updaters.iter().filter(|u| u.host() == name.host) {
+            let res = match updater.fetch_repository(&name) {
+                Ok(Some(repo)) => self.store_repository(conn, updater.host(), repo),
+                Err(err) => {
+                    warn!("failed to collect `{}` stats: {}", updater.host(), err);
+                    return Ok(None);
+                }
+                _ => continue,
+            };
+            return match res {
+                Ok(repo_id) => Ok(Some(repo_id)),
+                Err(err) => {
+                    warn!("failed to collect `{}` stats: {}", updater.host(), err);
+                    Ok(None)
+                }
+            };
+        }
+        // It means that none of our updaters have a matching host.
+        Ok(None)
+    }
+
+    pub fn update_all_crates(&self) -> Result<()> {
+        let mut conn = self.pool.get()?;
+        for updater in &self.updaters {
+            info!("started updating `{}` repositories stats", updater.host());
+
+            let needs_update = conn
+                .query(
+                    "SELECT host_id
+                     FROM repositories
+                     WHERE host = $1 AND updated_at < NOW() - INTERVAL '1 day';",
+                    &[&updater.host()],
+                )?
+                .into_iter()
+                .map(|row| row.get(0))
+                .collect::<Vec<String>>();
+
+            if needs_update.is_empty() {
+                info!(
+                    "no `{}` repositories stats needed to be updated",
+                    updater.host()
+                );
+                continue;
+            }
+            for chunk in needs_update.chunks(updater.chunk_size()) {
```

Comment:
You can avoid the collect if you use Itertools::chunks: https://docs.rs/itertools/0.10.0/itertools/trait.Itertools.html#method.chunks. Up to you if you want to change it - it's already in our build dependencies, but not our runtime dependencies.

Reply:
Let's maybe keep that for another PR. :)

Reply:
(putting a FIXME so it's not forgotten)